dockerfile updates

Josh-Walker-GM opened the pull request:
WIP changes to the dockerfile to address the output of the automated analysis

**Running list of changes**
1. Clean apt-get cache
   Removes ~20MB from images which use apt-get
2. I set the `DOCKER_CONTENT_TRUST=1` environment variable locally to remove the related advisory.
3. Switched to use the node user which is setup inside the node base image. 

Other changes:
1. Renames the targets to be consistent and follow a "side_function" convention so api_serve, web_build etc.

Josh-Walker-GM committed:
Clean apt-get cache

Josh-Walker-GM committed:
small lint related change to the analysis script

Josh-Walker-GM committed:
Changes to use non-root user as early as possible. Renames the targets for consistency.

I may have broken the caching (the --mount stuff) - I'm not too sure yet. Best to get an expert to look over this.

Josh-Walker-GM commented:
@rainkinz Would you mind taking a look over my changes here. I'm mainly concerned about the user change because that now means the place is littered with `--chown=node:node` but I don't really see mention anywhere online that this is a bad thing?

Trying to something like `RUN chown -R node:node node_modules` takes ages - so long I cancelled it and gave up. That made me want to run the yarn commands as the node user then. 

I also had to ensure I cleaned the cache after now using the node user to run the yarn commands otherwise all the packages were duplicated within the `/home/node/.yarn/berry/cache` folder. I suspect there are some new cached entries for prisma which are making the image a few MB larger now. I don't know if we should remove these or not.

rainkinz committed:
* Added back 'console' target for now so that we can migrate
* Use shell to start web side so that  env var is evaluated. See: 'https://docs.docker.com/engine/reference/builder/#cmd' for details

rainkinz committed:
Begin adding notes about running Redwood with docker compose in development

rainkinz commented:
@Josh-Walker-GM I think using `--chown=node:node` is okay. It is a bit noisy, but I'm not sure of a good alternative. I think some of the copy commands could be combined which might help. I'll try to take a closer look tomorrow. 

I've pushed up a branch off of your branch with a couple of updates I needed to get the project working. I put them in a PR here: https://github.com/jtoar/test-project-docker/pull/3 Feel free to ignore, but one thing that won't work in the current Dockerfile is the web server start command. We need to run the command in a shell to evaluate $API_HOST variable. Ping me if you want to discuss.

Josh-Walker-GM committed:
Small change to use chown and add a comment

rainkinz committed:
Merge pull request #3 from jtoar/bjg-docker-related-edits

@rainkinz docker related edits

Josh-Walker-GM commented:
I think these dockerfiles are getting good! The things that remain in my head before we get this out are:

1. We need to think about the "scripts" directory. Do these also get pulled in with the API side? I think we probably should.
2. We should probably refactor the compose to be "dev-time" so that it syncs the src's as volumes, run's dev, etc. I have the feeling this doesn't need a new target and can use "base".
3. (framework) Is this docker work going to wait for our changes to `api-server`/`web-server` bins so we can use them in the CMD's rather than the current `rw-server`.
   * This raises the question of passing variables/config around and how best we want to approach this.
   * We should ensure these new packages are added as prod dependencies during the setup command because by default they will only be available via core/cli. 
4. Migrations... I personally like the console and think it's a robust approach to being able to execute commands within the appropriate environment. We likely want to have easy answers for users to quickly migrate as they deploy to fly/render. 
   * I likely just need to look deeper into these deploy providers. They likely accept some sort of config/option for this sort of thing - like coherence does if I remember correctly.

and I have the following TODOs in my head for when we're about to integrate these into a setup command:
1. We need to have two options:
   1. One big dockerfile for people/places that support `--target` during build
   6. Two seperate dockerfiles for web/api for people/places that don't

_^ just putting my thoughts down in writing - no decisions being made and roast these as needed_

Josh-Walker-GM committed:
Add dev version of the dockerfile

Required changes to redwood.toml for it to work - networking and browser opening

Josh-Walker-GM commented:
I added a dev docker compose which mounts the project and runs `yarn rw dev`. I think we want to mount the node_modules differently - we want our node_modules to be from the container not the host - so I have attempted to do so. 

I had to make changed to the redwood.toml to get it to work. The `open = true` fails because it can't open. We should try/catch this in the CLI and print out a message rather than some initially scary stack-trace. The host option is a known thing and we really need to do a big audit of our serve/dev options and behaviour. 

To use this I ran:
```bash
docker compose -f docker-compose.dev.yml up --build --remove-orphans
```
and then to get a console to do things like migrations:
```bash
docker compose -f docker-compose.dev.yml run --rm console /bin/bash
```

(I pushed it to the wrong branch so had to revert and cherry-pick... anyway it's here now)

jtoar committed:
Merge branch 'aws' into dockle-related-edits

jtoar committed:
wip

jtoar committed:
add script for console

jtoar commented on .gitignore:
```diff
@@ -22,3 +22,5 @@ api/src/lib/generateGraphiQLHeader.*
 
 postgres
+
+docker-compose.override.yml
```
@Josh-Walker-GM is this an artifact or does this file get generated by docker compose or something like that?

Josh-Walker-GM replied:
@rainkinz Could you help answer this one? I think this was merged in from your PR to this one.

rainkinz replied:
`docker-compose.override.yml` is automatically merged with your `docker-compose.yml` when you do anything with docker-compose. You generally would put mounts to your local file system etc that are specific to you. For example I usually mount my `~/.aws` directory and history so that in the console it's all there. I can demo for you both if you want.

rainkinz replied:
Sorry, `docker-compose.override.yml` is manually created.

jtoar commented on Dockerfile:
```diff
@@ -1,155 +1,143 @@
-FROM node:18-slim as base
+# base
+# ------------------------------------------------
+FROM node:18-bookworm-slim as base
```
We're using bookworm, a "codename" for a version of a Debian LTS release, because we think it's a best practice to pin the LTS of the OS we're using like we do for Node.js

jtoar commented on Dockerfile:
```diff
@@ -1,155 +1,143 @@
 
-RUN --mount=type=cache,target=/root/.yarn/berry/cache \
-    --mount=type=cache,target=/root/.cache \
+RUN --mount=type=cache,target=/home/node/.yarn/berry/cache,uid=1000 \
+    --mount=type=cache,target=/home/node/.cache,uid=1000 \
```
`uid=1000` is the node user from the Node.js image

jtoar reviewed (approved):
Added a few comments and one question but going to merge so I can get going on updating the other providers
